rdrf/rdrf/dynamic_data.py

```diff
@@ -30,79 +24,102 @@ class KeyValueMissing(Exception):
     pass
 
 
-class MultisectionGridFSFileHandler(object):
-    def __init__(self,
-                 gridfs_filestore,
-                 registry_code,
-                 form_model,
-                 multisection_code,
-                 form_section_items,
-                 existing_nested_data,
-                 index_map):
-
-        self.original_data = deepcopy(existing_nested_data)
-        self.gridfs_filestore = gridfs_filestore
-        self.registry_code = registry_code
-        self.form_model = form_model
-        self.multisection_code = multisection_code
-        self.form_section_items = form_section_items # list of section dicts in this multisection
-        self.existing_nested_data = existing_nested_data
-        self.index_map = index_map  # if sections have been deleted,
+def find_sections(doc, form_name=None, section_code=None,
+                  formp=None, sectionp=None):
+    """
+    Iterates through form sections of a mongo doc.
+      form_name: Filter by form
+      section_code: Filter by code
+      formp: predicate function(form_dict, index)
+      sectionp: predicate function(section_dict, index)
+    """
+    formp = formp or (lambda f, i: True)
+    sectionp = sectionp or (lambda s, i: True)
+
+    for f, form in enumerate(doc.get("forms") or []):
+        if (form_name is None or
+            (form.get("name") == form_name and formp(form, f))):
+            for s, section in enumerate(form.get("sections") or []):
+                if (section_code is None or
+                    (section.get("code") == section_code and
+                     sectionp(section, s))):
+                    yield section
+
+def find_cdes(doc, form_name=None, section_code=None, cde_code=None,
```

Comment:
good idea  - I was going to do something similar in utils but good here.   Also I wonder if we could annotate these generators with contextlib.contextmanager